test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/RequestTests.cs

```diff
@@ -76,6 +71,65 @@ public class RequestTests
             }
         }
 
+        [ConditionalFact]
+        [FrameworkSkipCondition(RuntimeFrameworks.Mono, SkipReason = "Needs to run on 64-bit Mono, which isn't available on OSX.")]
+        public async Task LargeMultipartUpload()
+        {
+            var port = PortManager.GetPort();
+            var builder = new WebHostBuilder()
+                .UseServer("Microsoft.AspNetCore.Server.Kestrel")
+                .UseUrls($"http://localhost:{port}/")
+                .Configure(app =>
+                {
+                    app.Run(async context =>
+                    {
+                        long total = 0;
+                        var bytes = new byte[1024];
+                        var count = await context.Request.Body.ReadAsync(bytes, 0, bytes.Length);
+                        while (count > 0)
+                        {
+                            total += count;
+                            count = await context.Request.Body.ReadAsync(bytes, 0, bytes.Length);
+                        }
+                        await context.Response.WriteAsync(total.ToString(CultureInfo.InvariantCulture));
+                    });
+                });
+
+            using (var host = builder.Build())
+            {
+                host.Start();
+
+                using (var client = new HttpClient())
+                {
+                    using (var form = new MultipartFormDataContent())
+                    {
+                        const int oneGigabyte = 1024 * 1024 * 1024;
+                        const int files = 2;
+
+                        for (int i = 0; i < files; i++)
+                        {
+                            var fileName = Guid.NewGuid().ToString();
+                            var bytes = new byte[oneGigabyte];
+                            var content = new ByteArrayContent(bytes);
+                            content.Headers.ContentDisposition = new ContentDispositionHeaderValue("form-data")
+                            {
+                                Name = "\"files\"",
+                                FileName = "\"" + fileName + "\""
+                            };
+                            content.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
+
+                            form.Add(new ByteArrayContent(bytes), "file", fileName);
```

Comment:
BTW. I already chaged this test to use 2048 1MB files. This still had memory problems and this seemed mostly to be an issue with HttpClient. Most of the memory was still used up by tons of massive byte arrays rooted by `HttpRequestMessage`.

    0:000> !DumpObj /d 0000020f9b752a60
    Name:        System.Byte[]
    MethodTable: 00007ffbd1f27e00
    EEClass:     00007ffbd16d6c98
    Size:        1048600(0x100018) bytes
    Array:       Rank 1, Number of elements 1048576, Type Byte (Print Array)
    Content:     ................................................................................................................................
    Fields:
    None
    0:000> !gcroot 0000020f9b752a60
    Thread 41bc:
    *** ERROR: Module load completed but symbols could not be loaded for Microsoft.AspNetCore.Server.Kestrel.dll
        000000e68813e630 00007ffb790d8d0c Microsoft.AspNetCore.Server.Kestrel.KestrelThread.ThreadStart(System.Object)
            rbp-80: 000000e68813e680
                ->  0000020f00118fa8 System.Action`2[[System.Action`1[[System.IntPtr, mscorlib]], mscorlib],[System.IntPtr, mscorlib]]
                ->  0000020f000f41b8 Microsoft.AspNetCore.Server.Kestrel.KestrelThread
                ->  0000020f000f3680 Microsoft.AspNetCore.Server.Kestrel.KestrelEngine
                ->  0000020f000f3360 Microsoft.AspNetCore.Server.Kestrel.Http.DateHeaderValueManager
                ->  0000020f000f34b0 System.Threading.Timer
                ->  0000020f000f3520 System.Threading.TimerHolder
                ->  0000020f000f34c8 System.Threading.TimerQueueTimer
                ->  0000020f002eea18 System.Threading.TimerQueueTimer
                ->  0000020f002ee898 System.Threading.CancellationTokenSource+LinkedCancellationTokenSource
                ->  0000020f002ef8c8 System.Threading.SparselyPopulatedArray`1[[System.Threading.CancellationCallbackInfo, mscorlib]][]
                ->  0000020f002ef900 System.Threading.SparselyPopulatedArray`1[[System.Threading.CancellationCallbackInfo, mscorlib]]
                ->  0000020f002ef920 System.Threading.SparselyPopulatedArrayFragment`1[[System.Threading.CancellationCallbackInfo, mscorlib]]
                ->  0000020f002ef950 System.Threading.CancellationCallbackInfo[]
                ->  0000020f002ef898 System.Threading.CancellationCallbackInfo
                ->  0000020f002eeb98 System.Net.Http.WinHttpRequestState
                ->  0000020f002eeb30 System.Threading.Tasks.TaskCompletionSource`1[[System.Net.Http.HttpResponseMessage, System.Net.Http]]
                ->  0000020f002eeb48 System.Threading.Tasks.Task`1[[System.Net.Http.HttpResponseMessage, System.Net.Http]]
                ->  0000020f002eece8 System.Threading.Tasks.StandardTaskContinuation
                ->  0000020f002eec98 System.Threading.Tasks.ContinuationTaskFromResultTask`1[[System.Net.Http.HttpResponseMessage, System.Net.Http]]
                ->  0000020f002eeaf0 System.Action`1[[System.Threading.Tasks.Task`1[[System.Net.Http.HttpResponseMessage, System.Net.Http]], mscorlib]]
                ->  0000020f002ee860 System.Net.Http.HttpClient+<>c__DisplayClass55_0
                ->  0000020f002ee818 System.Net.Http.HttpRequestMessage
                ->  0000020f0011d8c0 System.Net.Http.MultipartFormDataContent
                ->  0000020f0011dda8 System.Collections.Generic.List`1[[System.Net.Http.HttpContent, System.Net.Http]]
                ->  0000020f001f8a90 System.Net.Http.HttpContent[]
                ->  0000020f002e9058 System.Net.Http.ByteArrayContent
                ->  0000020f9b752a60 System.Byte[]

    Found 1 unique roots (run '!GCRoot -all' to see all roots).

Ignore the DateHeaderValueManager in the GCRoot chain, that's caused by the way .NET timers use linked lists.

A quick search seems to indicate you can disable request buffering if you have direct access to an HttpWebReqeust by setting the AllowWriteStreamBuffering property to false. Unfortunately, this doesn't appear to be accessible through the HttpClient API. Does anyone here know how to disable request buffering for HttpClient?

If not, we can probably change this test to use a Socket to make the request and file an issue against HttpClient to support a way to disable request buffering.